server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -100,18 +98,24 @@ public void processExistingRecoveries(RoutingAllocation allocation) {
                 if (matchingNodes.getNodeWithHighestMatch() != null) {
                     DiscoveryNode currentNode = allocation.nodes().get(shard.currentNodeId());
                     DiscoveryNode nodeWithHighestMatch = matchingNodes.getNodeWithHighestMatch();
-                    // current node will not be in matchingNodes as it is filtered away by SameShardAllocationDecider
-                    final String currentSyncId;
-                    if (shardStores.getData().containsKey(currentNode)) {
-                        currentSyncId = shardStores.getData().get(currentNode).storeFilesMetaData().syncId();
-                    } else {
-                        currentSyncId = null;
-                    }
+                    BooleanSupplier currentNodeCanSkipPhase1 = () -> {
+                        TransportNodesListShardStoreMetaData.StoreFilesMetaData storeMetadata =
+                            shardStores.getData().get(currentNode).storeFilesMetaData();
+                        // current node will not be in matchingNodes as it is filtered away by SameShardAllocationDecider
+                        if (storeMetadata == null) {
+                            return false;
+                        }
+                        IndexMetaData indexMetadata = allocation.metaData().index(shard.index());
+                        if (canPerformOperationBasedRecovery(indexMetadata, primaryStore, currentNode, storeMetadata)) {
+                            return true;
```

Comment:
> has a PRRL that matches the primary

Did you mean that we should only cancel the current recovery if the replacement is a noop recovery?

Reply:
> we should only cancel the current recovery if the replacement is a noop 

Yes.

(we can't be sure it will be a no-op recovery by the time it runs, but that's the best we can do here).

Reply:
Ok, thanks David. With this, we do not need to treat closed/frozen indices specially.